### PR TITLE
added n-loops and tuple argument ability to @threads

### PR DESCRIPTION
**Pull request by marius311**
This lets you use `@threads` with n-loops, 

``` julia
a = zeros(5,5)
@threads for i=1:5, j=1:5
    a[i,j] = threadid()
end
a
5×5 Array{Float64,2}:
 1.0  6.0  3.0  8.0  5.0
 2.0  7.0  4.0  1.0  6.0
 3.0  8.0  5.0  2.0  7.0
 4.0  1.0  6.0  3.0  8.0
 5.0  2.0  7.0  4.0  1.0
```

and with tuple arguments, 

``` julia
a = zeros(5,5)
@threads for (i,j) in [(1,2),(3,4)]
    a[i,j] = threadid()
end
a
5×5 Array{Float64,2}:
 0.0  1.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  2.0  0.0
 0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0
```

(or with a mixture of both). 

Note from the first example that I changed the way the data is split among threads so that if `i` and `j` index an array, the threads move through the array in memory order so that caching is improved. 

Also, the old version declared the loop indices as `local` which I removed. To me it seems that if the user didn't do this, the `@threads` macro has no reason to do it for them. But please verify I haven't missed something (cc @ArchRobison who wrote that I believe). 

I think someone that knows Julia better than I could certainly improve my `ndi` function to make it faster, maybe move more of it to compile-time.

Anyway, this'll be my first contribution to the Julia language so please let me know if there's anything that needs fixing! Thanks.

### CODE REVIEW

**Comment by marius311** [comment]
See also https://github.com/JuliaLang/julia/issues/18263

File: base/threadingconstructs.jl [hunk 1]
```diff
@@ -18,39 +18,31 @@ on `threadid()`.
...
-                else
-                    f = f + rem
-                    l = l + rem
-                end
-            end
-            # run this thread's iterations
-            for i = f:l
-                local $(esc(lidx)) = Base.unsafe_getindex(r,i)
+            # each thread does every `nthreads`-th flattened index
+            for i = tid:nthreads():*(dims...)
```
Comment (timholy) [inline base/threadingconstructs.jl]: If I understand correctly, having 4 threads will cause thread 1 to handle item 1, 5, 9, etc. Compared to the current strategy, this seems likely to hurt cache efficiency. Have you benchmarked? 8 threads on `Float64` would be predicted to be a particularly bad case for this approach.

File: base/threadingconstructs.jl [hunk 2]
```diff
@@ -18,39 +18,31 @@ on `threadid()`.
...
+                $(Expr(:tuple,(Symbol("i_$i") for i=1:ndim)...)) = ndi(i,dims)
```
Comment (timholy) [inline base/threadingconstructs.jl]: This will also call `divrem` on each iteration. `divrem` is *really* slow.

**Comment by timholy** [comment]
The functionality is great, many thanks! And kudos to you for coming up with something that's more functional but fewer lines of code.

However, I think you need to find a way to partition the work without calling `divrem` for each element; something a bit more ahead-of-time, like the current strategy for a single range.

**Comment by yuyichao** [comment]
> Note from the first example that I changed the way the data is split among threads so that if `i` and `j` index an array, the threads move through the array in memory order so that caching is improved.

As @timholy said, this will probably cause worse cache behavior as different cores fighting for the same cachline.

> Also, the old version declared the loop indices as `local` which I removed.

The loop variable should never be shared so it should be `local`.

**Comment by timholy** [comment]
@marius311, I just realized there may be some confusion about storage order. Julia is "first index fastest" (aka column-major), and loops are traversed in last-iterator-fastest order. So write code like
```jl
for c = 1:col, r = 1:row
    s += A[r,c]
end
```
The order of every item in this example is carefully chosen, and almost anything else is "wrong."

**Comment by marius311** [comment]
Thanks both for taking a look.

Re: caching, quite possibly I don't fully understand how the caches work, but I thought that since the threads share the same cache, that it would be advantageous to have threads 1-8 do elements 1-8, then together move onto 8-16, then 16-24, etc... so that the relevant part of the array is in cache for all the threads most of the time. But I can change it back to the old behavior. 

Re: divrem, yea, that's a good point. Let me think about how to go about removing the need. 

Re: `local`, the loop variable seems to bind to a variable in the parent scope if it exists by default. Therefore with the current implementation you get the following unexpected behavior:

```julia
i=0
for i=1:2; end
@show i
i = 2
```

```julia
i=0
@threads for i=1:2; end
@show i
i = 0
```

Removing the `local` makes both of those give consistent results so I do think its the right thing to do.

**Comment by timholy** [comment]
See the accepted answer here: http://stackoverflow.com/questions/4802565/multiple-threads-and-cpu-cache. Some cores may be on entirely different chips/dies.

**Comment by yuyichao** [comment]
> I thought that since the threads share the same cache, that it would be advantageous to have threads 1-8 do elements 1-8, then together move onto 8-16, then 16-24, etc.

No. Different cores don't share all the caches. Doing the loop in this order will have really bad L1/L2 cacheline ping-pong between cores.

> Therefore with the current implementation you get the following unexpected behavior:

If `i` is in parent scope the behavior is always wrong for threaded loop and there's no way to avoid different behavior for threaded/non-threaded version when there's write to shared variable in the loop body. The current implementation is AFAIK the best way to avoid this error.

**Comment by timholy** [comment]
If you don't know about `CartesianRange` and `CartesianIndex`, those might help. But #9080 makes me sad (I have that issue number memorized).

**Comment by marius311** [comment]
> If i is in parent scope the behavior is always wrong for threaded loop and there's no way to avoid different behavior for threaded/non-threaded version when there's write to shared variable in the loop body. The current implementation is AFAIK the best way to avoid this error.

Sorry I'm not totally understanding. Lets call the current version of `@threads` in Julia `@threads`, and my proposed version which removes the `local` declaration `@threads2`. Then I have that `@threads` sometimes doesn't give the same behavior as an unparallelized for loop, but `@threads2` always does. Isn't that better? 

```julia

i=0
for i=1:2; end
@show i
> i = 2

i=0
@threads for i=1:2; end
@show i
> i = 0

i=0
@threads2 for i=1:2; end
@show i
> i = 2
```

This also just lets the user control if they want a `local` or not, unlike `@threads` (which effectively forces it for them),

```julia

i=0
for i=1:2; local i; end
@show i
> i = 0

i=0
@threads2 for i=1:2; local i; end
@show i
> i = 0
```

**Comment by yuyichao** [comment]
> but `@threads2` always does

This is not true. More importantly ref https://github.com/JuliaLang/julia/issues/14948 .

**Comment by marius311** [comment]
Ahhh right, duh, the different threads can't all be changing the same iteration variable at once. So yea, it has to be `local`, even if it might lead to the aforementioned difference with an unparallelized loop.

**Comment by marius311** [comment]
> See the accepted answer here: http://stackoverflow.com/questions/4802565/multiple-threads-and-cpu-cache. Some cores may be on entirely different chips/dies.

Thanks, yea, that explains it nicely. Ok, version 2 coming up where I don't get nearly every part of this PR wrong :)

**Comment by marius311** [comment]
Updated version adds back the declaration of loop indices as local. It also splits the data efficiently in the correct way, 

```julia
a = zeros(5,5)
@threads for c=1:5, r=1:5
    a[r,c] = threadid()
end
a
5×5 Array{Float64,2}:
 1.0  2.0  4.0  5.0  7.0
 1.0  2.0  4.0  6.0  7.0
 1.0  3.0  4.0  6.0  8.0
 1.0  3.0  5.0  6.0  8.0
 2.0  3.0  5.0  7.0  8.0
```

I shortened the code of the original partitioning logic, discovering `ind2sub` also helped. However, I don't see how one could make use of CartesianRange here, you'd need something like described [here](https://groups.google.com/forum/#!topic/julia-users/uFAqu31gHq4). Without something like that existing, I've basically hardcoded the "increment and carry logic" inside the loop. Maybe it can be done better? For now, it at least removes introducing the `divrem` in the loop.

**Comment by marius311** [comment]
Just a friendly ping for someone to take a look at this. I'm happy to make any other modifications necessary.

File: base/threadingconstructs.jl [hunk 3]
```diff
@@ -18,39 +18,52 @@ on `threadid()`.
...
+
+            # get N-d index (i1,i2,i3...) from linear index (and move back one
+            # since the first thing we do in the for loop is increment)
+            i_lidxs = [ind2sub(dims,start-1)...]
+
+            for i=1:len
+                # increment and carry N-d index (i1,i2,i3,...)
+                i_lidxs[1] += 1
+                for j=1:$ndim
+                    if i_lidxs[j]>dims[j]
```
Comment (marius311) [inline base/threadingconstructs.jl]: Thanks, yea, that's a much better. I will play around with benchmarking to see how this compares to both the original code and un@thread'ed.

**Comment by kpamnany** [comment]
I think this is useful+helpful functionality. I haven't reviewed it closely because... we'll be opening a PR "soon" (hopefully, shortly after the 0.6 release), for a new threading runtime that will integrate Julia tasks, remove the current interface (`jl_threading_run()`), and instead provide a Cilk-like interface (spawn, sync, and parfor). This will have to be reworked when that comes in, unfortunately.

I have no objection to merging this now anyway (esp. since Arch has already had a look), so that there's a template for this functionality.

File: base/threadingconstructs.jl (hunk 3 above)
Comment (kpamnany) [inline base/threadingconstructs.jl]: Any data on performance impact, if any?

**Comment by StefanKarpinski** [comment]
That's super exciting, Kiran! I would hold off until your big threading PR is ready, however. Right before a feature freeze is not the time for this sort of change. Right after is a different story!

**Comment by marius311** [comment]
Fyi, a while back I benchmarked the improved increment and carry suggestion from @archrobinson from above, and it is indeed a somewhat significant slowdown for the 1d case (I forget the exact number now) as compared to the original code. At that point I kind of dropped it. 

Maybe I should split off the tuple argument part into a separate PR since that seems fairly straight forward, and let a more Julia expert work on the multiple loops?

**Comment by ChrisRackauckas** [comment]
It would be nice for this to be revived. `@threads` ends up still being a nice thing to have around.

**Comment by marius311** [comment]
Tuple argments work now, and the approach for n-loops in this PR was probably wrong, but anyone is welcome to steal any code here for a fresh stab at this should they want to.

**Pull request closed without merging**